satpy/modifiers/_crefl_utils.py

```diff
@@ -49,6 +88,434 @@
+
+    def __call__(self, sensor_azimuth, sensor_zenith, solar_azimuth, solar_zenith, avg_elevation):
+        refl = self._refl
+        height = self._height_from_avg_elevation(avg_elevation)
+        coeffs_helper = self.coeffs_cls(refl.attrs["wavelength"], refl.attrs["resolution"])
+        coeffs = coeffs_helper()
+        mus = np.cos(np.deg2rad(solar_zenith))
+        mus = mus.where(mus >= 0)
+        muv = np.cos(np.deg2rad(sensor_zenith))
+        phi = solar_azimuth - sensor_azimuth
+        corr_refl = self._run_crefl(mus, muv, phi, solar_zenith, sensor_zenith, height, coeffs)
+        if self._is_percent:
+            corr_refl = corr_refl * 100.0
+        return xr.DataArray(corr_refl, dims=refl.dims, coords=refl.coords, attrs=refl.attrs)
+
+    def _run_crefl(self, mus, muv, phi, solar_zenith, sensor_zenith, height, coeffs):
+        raise NotImplementedError()
+
+    def _height_from_avg_elevation(self, avg_elevation: Optional[np.ndarray]) -> da.Array:
+        """Get digital elevation map data for our granule with ocean fill value set to 0."""
+        if avg_elevation is None:
+            LOG.debug("No average elevation information provided in CREFL")
+            # height = np.zeros(lon.shape, dtype=np.float64)
+            height = 0.
+        else:
+            LOG.debug("Using average elevation information provided to CREFL")
+            lon, lat = self._refl.attrs['area'].get_lonlats(chunks=self._refl.chunks)
+            height = da.map_blocks(_space_mask_height, lon, lat, avg_elevation,
+                                   chunks=lon.chunks, dtype=avg_elevation.dtype)
+        return height
+
+
+class _ABICREFLRunner(_CREFLRunner):
+    @property
+    def coeffs_cls(self) -> Type[_Coefficients]:
+        return _ABICoefficients
+
+    def _run_crefl(self, mus, muv, phi, solar_zenith, sensor_zenith, height, coeffs):
+        LOG.debug("Using ABI CREFL algorithm")
+        return da.map_blocks(_run_crefl_abi, self._refl.data, mus.data, muv.data, phi.data,
+                             solar_zenith.data, sensor_zenith.data, height, *coeffs,
+                             meta=np.ndarray((), dtype=self._refl.dtype),
+                             chunks=self._refl.chunks, dtype=self._refl.dtype,
+                             )
+
+
+class _VIIRSCREFLRunner(_CREFLRunner):
+    @property
+    def coeffs_cls(self) -> Type[_Coefficients]:
+        return _VIIRSCoefficients
+
+    def _run_crefl(self, mus, muv, phi, solar_zenith, sensor_zenith, height, coeffs):
+        LOG.debug("Using VIIRS CREFL algorithm")
+        return da.map_blocks(_run_crefl, self._refl.data, mus.data, muv.data, phi.data,
+                             height, self._refl.attrs.get("sensor"), *coeffs,
+                             meta=np.ndarray((), dtype=self._refl.dtype),
+                             chunks=self._refl.chunks, dtype=self._refl.dtype,
+                             )
+
+
+class _MODISCREFLRunner(_VIIRSCREFLRunner):
```

Comment:
There should be a common base class for the viirs and modis variants, at least because the debug message would be wrong in the modis crefl runner.

Reply:
Good catch. I see a couple options, which one below do you think is cleanest? This assumes that the problem you have is that the log message is wrong not MODIS and VIIRS should share a base class versus MODIS subclassing VIIRS.

1. I make the runner base class have a "name" class attribute that is used in the debug log message to say which runner is running.
2. I remove the log message.
3. Create a separate class for MODIS (based off of the _CREFLRunner base class) that duplicates the `map_blocks` call but with the correct log message.
4. Have a VIIRS/MODIS base class with a `_run_crefl` method that just calls another method to run map_blocks. The VIIRS and MODIS subclasses then override this method adding a single debug message.
5. Move the log message to the `__init__`.

Reply:
I was thinking about 4.

Reply:
Ended up finding another way using `super()`. Let me know what you think.